lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs

```diff
@@ -0,0 +1,196 @@
+    /// of times this metric has been updated. If the metric is immutable, it keeps a record of updates.
+    /// Once the data has been processed, the records and count will reset.
+    /// </summary>
+    [JsonObject]
+    public sealed class MetricTracker : IObserver<IMetric>
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(MetricsData));
+
+        private bool _keepUpdateHistory;
+
+        [JsonProperty]
+        private IMetric _mirror;
+
+        /// <summary>
+        /// List of the history of values this metric has held. If _keepUpdateHistory is false, only holds current value.
+        /// </summary>
+        [JsonProperty]
+        private IList<MetricRecord> _records;
+
+        /// <summary>
+        /// Number of times metric has been updated since last processed.
+        /// </summary>
+        [JsonProperty]
+        internal int ChangesSinceLastSink;
+
+        /// <summary>
+        /// Constructor for metricData
+        /// </summary>
+        /// <param name="metric"></param>
+        /// <param name="initialValue"></param>
+        internal MetricTracker(IMetric metric)
+        {
+            Subscribe(metric);
+            _mirror = metric;
+            ChangesSinceLastSink = 0;
+            _keepUpdateHistory = metric.IsImmutable;
+            _records = new List<MetricRecord>();
+            _records.Add(CreateMetricRecord());
+        }
+
+        [JsonConstructor]
+        internal MetricTracker(IList<MetricRecord> records, int changes)
+        {
+            _records = records;
+            ChangesSinceLastSink = changes;
+        }
+
+        /// <summary>
+        /// Reset records.
+        /// </summary>
+        internal void ResetChangesSinceLastSink()
+        {
+            ChangesSinceLastSink = 0;
+            _records.Clear();
+        }
+
+        /// <summary>
+        /// When new metric data is received, update the value and records so it reflects the new data.
+        /// </summary>
+        /// <param name="metric">Metric data received.</param>
+        internal MetricTracker UpdateMetric(MetricTracker metric)
+        {
+            _mirror = metric.GetMetric();
+            if (metric.ChangesSinceLastSink > 0)
+            {
+                if (_keepUpdateHistory)
+                {
+                    _records.Concat(metric._records);
+                }
+                else
+                {
+                    _records = metric.GetMetricRecords().ToList();
+                }
+            }
+            ChangesSinceLastSink += metric.ChangesSinceLastSink;
+            return this;
+        }
+
+        /// <summary>
+        /// Updates metric value with metric object received.
+        /// </summary>
+        /// <param name="me">New metric.</param>
+        internal void UpdateMetric(IMetric me)
+        {
+            ChangesSinceLastSink++;
+            _mirror = me;
+            UpdateRecords();
+        }
+
+        private void UpdateRecords()
```

Comment:
If possible, try to stay away from internal states and take a functional programming approach. If you isolate all your logic to the function you have less of a chance for unwanted side-effects.
```csharp
private void UpdateRecords(IMetric metric)
{
    var newRecord = CreateMetricRecord(metric);
```